request-authentication.md

```diff
@@ -6,13 +6,85 @@ pagination_next: null
 unique secret received when the webhook was registered. The authorization is a
 sha256 HMAC hash of the request date, content, and URI that is created using the
 webhook secret.
 
+<Tabs
+    defaultValue="js" 
+    groupId="example-choice" 
+    values={[ {label: 'JavaScript', value: 'js'}, {label: '.Net C#', value: 'csharp'}, ]}>
+
+<TabItem value="js">
+
+To run: `node webhook.mjs`
+```js
+import crypto from "crypto";
+
+// The secret is found in the response body when you register a webhook. 
+// Note: The secret is unique per hook registration (EventType+YourURL).
+const secret = 'A0+AeKBRG2KRGvnNwJpQlb6IJFk48CKXCIcrLoHncVJKDILsQSxS6NWCccwWm6r6FhGKhiHTBsG2wo/xU6FY/A=='
+
+// The complete url of the webhook, including query parameters if applicable. I.e. https://yoursite.com/webhooks/vmp/v1
+const requestUrl = "https://webhook.site/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63"
+// Part of the HTTP request header from Vipps MobilePay, found in the field 'x-ms-date'
+const dateTime = "Thu, 30 Mar 2023 08:38:32 GMT"
+
+// The signature/authorization sent by Vipps MobilePay for you to verify against. 
+// Part of the HTTP request header from Vipps MobilePay, found in the field 'authorization'
+const actualSignature = "HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U="
+
+// HTTP Body
+const payloadJson = {
+  "some-unique-content":"ee6e441b-cc4a-46f8-895d-a5af79bcc233/hello-world"
+}
+
+const hashedPayload =  crypto.createHash("sha256").update(JSON.stringify(payloadJson)).digest("base64")
+// returns 'lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4='
+
+// The complete url of the call to the webhook, including query parameters if applicable
+const url = new URL(requestUrl)
+const pathAndQuery = url.pathname + url.search; 
+// returns '/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63'
+const host = url.host 
+// returns 'webhook.site'
+
+// Example is using HTTP POST, hence starting with 'POST\n'
+const expectedSignedString = `POST\n${pathAndQuery}\n${dateTime};${host};${hashedPayload}`
+/* returns:
+POST
+/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63
+Thu, 30 Mar 2023 08:38:32 GMT;webhook.site;lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4=
+*/
+
+// Run the HMAC algorithm on the expectedSignedString with the secret for the hook
+const expectedSignature = crypto.createHmac("sha256", secret)
+  .update(expectedSignedString)
+  .digest("base64")
+// returns 'agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U=' 
+// which matches the expected signature
+
+const expectedAuth = `HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=${expectedSignature}`
+
+if(expectedAuth !== actualSignature) {
+  // Log the error.
+  // Fail, stop processing
+  throw new Error('Authorization fields do not match expected value!');
+}
+// Continue with business logic related to the webhook.
+console.log('Success! Business logic awaits!')
```

Comment:
I think it's nice to formulate this as a working example, 
but I think you can take it a step further and start at 
the top with an actual node http request. 

E.g. define a method `isValidWebhookCallback(request)` that takes a request
and returns true or false depending on whether the request was valid.

Then create a dummy request object and use the method
```js
const isValidWebhookRequest = (request) => {
    ...
}

const incommingRequest = ...
assert.strictEqual(isValidWebhookRequest(incommingRequest), true)
```

Reply:
Good idea, I initially created a remix.run app+ngrok where I tested this and some other stuff.
So I just pulled out the relevant code and tested that I could run it on the command line.

But this is a good idea "going forward" 😁

Reply:
```javascript
import crypto from "crypto";

function validate(url, dateTime, signature, secret, body) {

  const pathAndQuery = url.pathname + url.search; 
  // '/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63'

  const host = url.host 
  // 'webhook.site'

  const hashedPayload =  crypto
    .createHash("sha256")
    .update(JSON.stringify(httpBody))
    .digest("base64") 
    // evaluates to 'lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4='

  const expectedSignedString = `POST\n${pathAndQuery}\n${dateTime};${host};${hashedPayload}`
  // evaluates to:
  // POST
  // /e2cee29b-012e-4f1d-8ef4-e95fd74a7a63
  // Thu, 30 Mar 2023 08:38:32 GMT;webhook.site;lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4=
  
  // Run the HMAC SHA256 algorithm on the expectedSignedString with the secret for the hook
  const expectedSignature = crypto
    .createHmac("sha256", secret)
    .update(expectedSignedString)
    .digest("base64")
    // evaluates to: 'agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U='

  const validSignature = `HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=${expectedSignature}`

  return validSignature !== actualSignature;
}




// The secret is found in the response body when you register a webhook.
// Note: The secret is unique per hook registration, which may cover multiple event types.
const secret = 'A0+AeKBRG2KRGvnNwJpQlb6IJFk48CKXCIcrLoHncVJKDILsQSxS6NWCccwWm6r6FhGKhiHTBsG2wo/xU6FY/A=='

const requestUrl = new URL("https://webhook.site/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63")

// Part of the HTTP request header from Vipps MobilePay, found in the field 'x-ms-date'
const dateTime = "Thu, 30 Mar 2023 08:38:32 GMT"

// The signature/authorization sent by Vipps MobilePay for you to verify against. 
// Part of the HTTP request header from Vipps MobilePay, found in the field 'authorization'
const actualSignature = "HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U="

const httpBody = { "some-unique-content":"ee6e441b-cc4a-46f8-895d-a5af79bcc233/hello-world" }

const isThisAValidWebhookRequest = validate(requestUrl, dateTime, actualSignature, secret, httpBody)
```